jwk/helper.go

```diff
@@ -38,71 +38,41 @@ import (
 	jose "gopkg.in/square/go-jose.v2"
 )
 
-func EnsureAsymmetricKeypairExists(ctx context.Context, r InternalRegistry, g KeyGenerator, set string) error {
-	_, _, err := AsymmetricKeypair(ctx, r, g, set)
+func EnsureAsymmetricKeypairExists(ctx context.Context, r InternalRegistry, alg, set string) error {
+	_, _, err := GetOrGenerateKeys(ctx, r, r.KeyManager(), set, set, alg)
 	return err
 }
 
-func AsymmetricKeypair(ctx context.Context, r InternalRegistry, g KeyGenerator, set string) (public, private *jose.JSONWebKey, err error) {
-	priv, err := GetOrCreateKey(ctx, r, g, set, "private")
-	if err != nil {
-		return nil, nil, err
-	}
-
-	pub, err := GetOrCreateKey(ctx, r, g, set, "public")
-	if err != nil {
-		return nil, nil, err
-	}
-
-	return pub, priv, nil
-}
-
-func GetOrCreateKey(ctx context.Context, r InternalRegistry, g KeyGenerator, set, prefix string) (*jose.JSONWebKey, error) {
-	keys, err := r.KeyManager().GetKeySet(ctx, set)
+func GetOrGenerateKeys(ctx context.Context, r InternalRegistry, m Manager, set, kid, alg string) (public, private *jose.JSONWebKey, err error) {
+	keys, err := m.GetKeySet(ctx, set)
 	if errors.Is(err, x.ErrNotFound) || keys != nil && len(keys.Keys) == 0 {
 		r.Logger().Warnf("JSON Web Key Set \"%s\" does not exist yet, generating new key pair...", set)
-		keys, err = createKey(ctx, r, g, set)
+		keys, err = m.GenerateKeySet(ctx, set, kid, alg, "sig")
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	} else if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	key, err := FindKeyByPrefix(keys, prefix)
-	if err != nil {
-		r.Logger().Warnf("JSON Web Key with prefix %s not found in JSON Web Key Set %s, generating new key pair...", prefix, set)
-
-		keys, err = createKey(ctx, r, g, set)
-		if err != nil {
-			return nil, err
+	pubKey, pubKeyErr := FindPublicKey(keys)
+	privKey, privKeyErr := FindPrivateKey(keys)
+	if pubKeyErr == nil && privKeyErr == nil {
+		return pubKey, privKey, nil
+	} else {
+		if pubKeyErr != nil {
+			r.Logger().Warnf("Public JSON Web Key not found in JSON Web Key Set %s, generating new key pair...", set)
+		} else {
+			r.Logger().Warnf("Private JSON Web Key not found in JSON Web Key Set %s, generating new key pair...", set)
 		}
-
-		key, err = FindKeyByPrefix(keys, prefix)
+		keys, err = m.GenerateKeySet(ctx, set, kid, alg, "sig")
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
+		pubKey, _ = FindPublicKey(keys)
+		privKey, _ = FindPrivateKey(keys)
+		return pubKey, privKey, nil
```

Comment:
Are we sure the errors from Find* can never be nil? I think we should return them if they do happen, even if it's unlikely, so e.g.:

```suggestion
		pubKey, err := FindPublicKey(keys)
		if err != nil {
			return nil, nil, err
		}
		privKey, err := FindPrivateKey(keys)
		if err != nil {
			return nil, nil, err
		}
		return pubKey, privKey, nil
```